### PR TITLE
haskellPackages.llvm-ffi: set LLVM

### PR DESCRIPTION
**Pull request by MaxHearnden**
###### Description of changes

Overrides LLVM from null to llvmPackages_13.libllvm

Fixes https://github.com/NixOS/nixpkgs/issues/227412

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thielema** [comment]
Must `llvm-ffi` be removed from `broken.yaml`, too?

**Comment by maralorn** [comment]
Yeah, that’s correct please remove the package from `broken.yaml`, because `hackage-packages.nix` is an auto generated file. Also `hydraPlatforms = lib.platforms.none;` line above the broken line should also be deleted.

I am not 100% certain about the LLVM override. @sternenseemann is that okay this way?

**Comment by MaxHearnden** [comment]
Just managed to properly unmark llvm-ffi as broken.

In case anyone else wants to unmark a package as broken the steps are

1. update pkgs/development/haskell-modules/configuration-hackage2nix/broken.yaml to remove the fixed package
2. run `maintainers/scripts/haskell/regenerate-transitive-broken-packages.sh` to update transitive-broken.yaml
3. run `maintainers/scripts/haskell/regenerate-hackage-packages.sh` to update hackage-packages.nix

The documented method in HACKING.md does not cover removing packages from broken.yaml

**Comment by maralorn** [comment]
@MaxHearnden that’s actually not quite right.

Running `regenerate-transitive-broken-packages.sh` does not look into the `broken.yaml` at all but uses the generated `hackage-packages.nix` file. Thus, if you only change `broken.yaml` and then run `regenerate-transitive-….sh`, the changes to `broken.yaml` won’t be reflected.

Also, the `transitive-broken.yaml` only affects whether we build packages on hydra, not if they are marked broken, so in general it’s not very urgent to update it, and we always do it before merging `haskell-updates` into `master`. So we generally don’t expect people to run it in their PRs, but it’s nice if you do.

So the correct steps are:

1.  update pkgs/development/haskell-modules/configuration-hackage2nix/broken.yaml to remove the fixed packageR
2. Run `maintainers/scripts/haskell/regenerate-hackage-packages.sh` to update `hackage-packages.nix`

and then optionally if you want to go the extra mile and see the effect for all dependencies on hydra as soon as possible:

3. run `maintainers/scripts/haskell/regenerate-transitive-broken-packages.sh` to update transitive-broken.yaml
4. run `maintainers/scripts/haskell/regenerate-hackage-packages.sh` again

**Comment by MaxHearnden** [comment]
I have hopefully done it correctly this time

**Comment by sternenseemann** [comment]
It is preferred to rebase PR branches on top of the target instead of merging the target branch back in again.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -2621,4 +2621,6 @@ self: super: {
   # compatability with other gi- packages.
   # Take another look when gi-webkit2 updates as it may have become compatible with libsoup-3
   gi-soup = assert versions.major self.gi-webkit2.version == "4"; self.gi-soup_2_4_28;
+
+  llvm-ffi = super.llvm-ffi.override {LLVM = pkgs.llvmPackages_13.libllvm;};
```
Comment (sternenseemann) [inline pkgs/development/haskell-modules/configuration-common.nix:2625]: Does it more sense to use the default LLVM (i.e. `self.llvmPackages.libllvm`) used for the compiler (maybe just to reduce closure size in some cases)? Seems like llvm-ffi supports multiple versions.

**Comment by thielema** [comment]
On Sat, 22 Apr 2023, sternenseemann wrote:

> @sternenseemann commented on this pull request.
> 
> ____________________________________________________________________________________________________
> 
> In pkgs/development/haskell-modules/configuration-common.nix:
> 
> > @@ -2621,4 +2621,6 @@ self: super: {
>    # compatability with other gi- packages.
>    # Take another look when gi-webkit2 updates as it may have become compatible with libsoup-3
>    gi-soup = assert versions.major self.gi-webkit2.version == "4"; self.gi-soup_2_4_28;
> +
> +  llvm-ffi = super.llvm-ffi.override {LLVM = pkgs.llvmPackages_13.libllvm;};
> 
> Does it more sense to use the default LLVM (i.e. self.llvmPackages.libllvm) used for the compiler
> (maybe just to reduce closure size in some cases)? Seems like llvm-ffi supports multiple versions.

I cannot promise that I always support the version that is chosen by Nix 
as default for GHC. What versions of GHC and LLVM have you chosen for the 
next NixOS release?

**Comment by maralorn** [comment]
We currently use llvm 12, and I don’t expect that to change before branch-off.

**Comment by thielema** [comment]
On Sun, 23 Apr 2023, maralorn wrote:

> We currently use llvm 12, and I don’t expect that to change before branch-off.

Then llvm-ffi must be built with Cabal flag "llvm1200" enabled.

**Comment by sternenseemann** [comment]
GHC chooses a range of supported LLVM versions and we have different ones per package set as a result. Since the newer GHC versions already support LLVM 14, it seems reasonable to assume that llvm-ffi doesn't keep up, so maybe pinning the version is the best course of action. Worst case is that you have to download LLVM twice when building `haskellPackages.llvm-ffi` on e.g. aarch64-linux.

**Comment by MaxHearnden** [comment]
I have, regenerated haskell packages using the updated script, set llvm to llvm 12 and enabled the cabal flag.

**Comment by maralorn** [comment]
But if we are pinning for the convenience of llvm-ffi anyway, why not leave it at pinning llvm 13?

**Comment by MaxHearnden** [comment]
I have simplified it to only using llvm 13 although pinning on llvm 12 would ensure that it doesn't break when llvm-ffi updates to llvm 14

**Comment by maralorn** [comment]
Fair point, but then we notice the build error and can update the pin to llvm 14 so that’s basically what we want.

Thank you for bearing with us!

**Comment by thielema** [comment]
I have updated `llvm-ffi` and `llvm-tf` to support LLVM-9 to LLVM-16. What version will be the default for the next NixOS release?

**Comment by maralorn** [comment]
@thielema Apparently LLVM 16, if I understand this correctly. https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/37

**Pull request merged**